src/components/Identity/createComponent.js

```diff
@@ -23,17 +25,78 @@ export default (processIdSyncs, config, logger, consent, eventManager) => {
     orgId,
     IDENTITY_COOKIE_KEY
   );
+  let deferredForOptInOld;
+  let deferredForVisitorECID;
   let deferredForIdentityCookie;
   const migration = createMigration({ orgId, consent });
   const hasIdentityCookie = () => Boolean(cookieJar.get(identityCookieName));
   const customerIds = createCustomerIds({ eventManager, consent, logger });
+  const Visitor = window.Visitor;
+  const doesVisitorExist =
+    isFunction(Visitor) && isFunction(Visitor.getInstance);
 
   // TO-DOCUMENT: We wait for ECID before trigger any events.
   const accommodateIdentityOnRequest = payload => {
     let promise;
 
+    const getVisitorECID = () => {
+      logger.log(
+        "Delaying request while using Visitor to retrieve ECID from server."
+      );
+
+      const visitor = Visitor.getInstance(orgId, {});
+
+      visitor.getMarketingCloudVisitorID(ecid => {
+        if (!ecid) {
+          return;
+        }
+
+        payload.expectResponse();
+        addEcidToPayload(payload, ecid);
+
+        if (
+          config.thirdPartyCookiesEnabled &&
+          areThirdPartyCookiesSupportedByDefault(getBrowser(window))
+        ) {
+          payload.useIdThirdPartyDomain();
+        }
+
+        logger.log(
+          "Resuming previously delayed request that was waiting for ECID from Visitor."
+        );
+
+        const deferred = deferredForOptInOld || deferredForVisitorECID;
+
+        deferred.resolve();
+      }, true);
+    };
+
     if (!hasIdentityCookie()) {
-      if (deferredForIdentityCookie) {
+      if (doesVisitorExist && !deferredForOptInOld && !deferredForVisitorECID) {
+        if (isObject(window.adobe) && isObject(window.adobe.optIn)) {
+          const optInOld = window.adobe.optIn;
+
+          deferredForOptInOld = defer();
```

Comment:
I would recommend not using `defer` for this case. Use a regularly constructed Promise where possible. In Aaron's Opinion™, a deferred should only be used when the thing resolving the promise is a task that the creator of the promise did not initiate. This is usually rare, resulting in deferreds often being abused and causing more problems than they try to solve (in fact, for a while you could create a deferred natively in some browsers until it was made obsolete).

Here, this code that's creating the promise is also the initiator of `optInOld.fetchPermissions`, so constructing a promise normally is most appropriate. I think you'll find your code to be simpler. I would start by breaking down each step in this process (1. fetching permissions and 2. Getting the visitor ID) into separate functions, each returning their own promise. Something like this:

```
const awaitVisitorOptIn = () => {
  return new Promise(resolve => {
    if (isObject(window.adobe) && isObject(window.adobe.optIn)) {
      const optInOld = window.adobe.optIn;
      optInOld.fetchPermissions(() => {
        if (optInOld.isApproved([optInOld.Categories.ECID])) {
          logger.log(
            "Received legacy opt in approval to let Visitor retrieve ECID from server."
          );

          resolve();
        }
      }, true);
      logger.log(
        "Delaying request while waiting for legacy opt in to let Visitor retrieve ECID from server."
      );
    } else {
      resolve();
    }
  });
};

const getVisitorECID = () => {
  return awaitVisitorOptIn().then(() => {
    return new Promise(resolve => {
      logger.log(
        "Delaying request while using Visitor to retrieve ECID from server."
      );

      const visitor = Visitor.getInstance(orgId, {});

      visitor.getMarketingCloudVisitorID(ecid => {
        if (!ecid) {
          return;
        }

        payload.expectResponse();
        addEcidToPayload(payload, ecid);

        if (
          config.thirdPartyCookiesEnabled &&
          areThirdPartyCookiesSupportedByDefault(getBrowser(window))
        ) {
          payload.useIdThirdPartyDomain();
        }

        logger.log(
          "Resuming previously delayed request that was waiting for ECID from Visitor."
        );

        resolve();
      }, true);
    });
  });
};
```

And put the functions inside the migration module. Then, like @jonsnyder said (and @jfkhoury eluded to), change `getEcidFromLegacy` to return promise. Then make `getEcidFromLegacy` leverage the two functions above. Let me know if that doesn't make sense.